### ISSUE TITLE
JWT authentication on websocket endpoints

### ISSUE THREAD

**Issue by bgins** [ereadingtool/ereadingtool#160]
The authentication mechanism used on the REST endpoints will not work for the websocket endpoints. The problem is that websockets do not allow you to set custom headers like the `Authorization` header that we use on our REST endpoints.

Currently, when setting up a websocket connection on an endpoint like `ws://localhost:8000/student/text_read/18/`, the backend fails with a `user.models.ReaderUser.student.RelatedObjectDoesNotExist: ReaderUser has no student` error. I think this may be because we are not authenticating the user, but I am not certain about that.

If that assumption is correct, we should find a way to authenticate users when we set up a websocket connection. There are a few ways we might do this:

- Use the WebSocket protocol header `Sec-WebSocket-Protocol` to pass the JWT
- Force the JWT onto a cookie client side and have the server look for it
- Append the JWT as a querysrting param on the websocket endpoint

Here is a discussion for reference: https://stackoverflow.com/questions/4361173/http-headers-in-websockets-client-api

**Comment by jdelamare** [comment]
Based on the need for `self.scope` to be accessible by both `flashcards/` as well as `text/`, I propose another project level directory. That would allow a clear and structured way for these consumers to access our custom `QueryAuthMiddleware`. The idea here will be borrowed and embellished from its original source [here](https://channels.readthedocs.io/en/latest/topics/authentication.html#custom-authentication). 

This directory should be called `auth` where a base class (maybe `CustomAuthMiddleware`?) that may be extended for future use will reside. Currently, the only subclass worth inheriting from `CustomAuthMiddleware` would be for authorizing producers to the websocket endpoints. Perhaps instead of `QueryAuthMiddleware` as the docs have provided we could go with `ProducerAuthMiddleware`. More research will need to be done.

I'll implement these locally and push upstream when something rudimentary works, perhaps under the branch name `upgrade/auth-middleware`.

**Comment by jdelamare** [comment]
Leaving notes for myself so I have a target and reference in the future. Looking to do away with `AuthMiddlewareStack` as it relies on Django's auth mechanism which is session based. As in, it would do a lookup on the backend. Since we're moving towards using a JWT this is unnecessary but will warrant the creation of the above `CustomAuthMiddleware` which will take the `URLRouter` as an argument as seen below.
 
https://github.com/ereadingtool/ereadingtool/blob/8fd9bbe6fb4bf6a43a92430ec70dff4b6cbb41de/ereadingtool/routing.py#L12-L23

**Comment by jdelamare** [comment]
There's now a new branch `upgrade/jwt_auth` which contains much of the logic necessary to give a user the go ahead to get a text when they provide a JWT. Unfortunately, I'm blocked by not having redis setup on my machine. That's next on the list of things to do. Which may bleed over to the DevOps stuff if we choose to microservice that redis instance out of the API's container.

**Comment by jdelamare** [comment]
Had another issue that was identical to this one, just auth on the websocket endpoints. Still not sure how we're handling unauthorized calls to the API. For a different issue / explanation.
